ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/paint/PaintFragment.kt

```diff
@@ -0,0 +1,187 @@
+
+        view.findViewById<View>(R.id.back_to_main).setOnClickListener(this)
+
+        val eraserView = view.findViewById<LinearLayout>(R.id.eraser_btn)
+        eraserView.setOnClickListener(this)
+        imageEraser = eraserView.findViewById(R.id.eraser_icon)
+
+        val brushView = view.findViewById<LinearLayout>(R.id.brush_btn)
+        brushView.setOnClickListener(this)
+        imageBrush = brushView.findViewById(R.id.brush_icon)
+
+        view.findViewById<View>(R.id.settings).setOnClickListener(this)
+    }
+
+    override fun onClick(view: View) {
+        if (view.id == R.id.back_to_main) {
+            backToMain()
+        } else if (view.id == R.id.eraser_btn) {
+            if (viewModel.isEraser.value == false) {
+                viewModel.setIsEraser(true)
+                toggleButtons()
+            }
+        } else if (view.id == R.id.brush_btn) {
+            if (viewModel.isEraser.value == true) {
+                viewModel.setIsEraser(false)
+                toggleButtons()
+            }
+        } else if (view.id == R.id.settings) {
+
+            if (viewModel.isEraser.value == true) {
+                showBottomSheet(EraserConfigDialog())
+            } else {
+                showBottomSheet(BrushConfigDialog())
+            }
+        }
+    }
+
+    private fun showBottomSheet(dialogFragment: BottomSheetDialogFragment) {
+        if (dialogFragment.isAdded) return
+        dialogFragment.show(parentFragmentManager, dialogFragment.tag)
+    }
+
+    override fun backToMain() {
+        activity.bottomGallery.currentItem = MainMenuFragment.INDEX
+        activity.mainImage.visibility = VISIBLE
+        activity.paintView.reset()
+        activity.paintView.visibility = View.GONE
+        activity.mode = EditImageActivity.MODE_NONE
+        activity.bannerFlipper.showPrevious()
+
+        viewModel.resetToDefault()
+    }
+
+    override fun onShow() {
+        activity.mode = MODE_PAINT
+        activity.mainImage.setImageBitmap(activity.mainBit)
+        activity.paintView.visibility = VISIBLE
+        activity.bannerFlipper.showNext()
+    }
+
+    private fun toggleButtons() {
+        imageEraser?.setImageResource(
+                if (viewModel.isEraser.value == true) R.drawable.ic_eraser_enabled
+                else R.drawable.ic_eraser_disabled)
+        imageBrush?.setImageResource(
+                if (viewModel.isEraser.value == true) R.drawable.ic_brush_grey_24dp
+                else R.drawable.ic_brush_white_24dp)
+    }
+
+    fun savePaintImage() {
+        compositeDisposable.clear()
+        val applyPaintDisposable = applyPaint(activity.mainBit)
+                .flatMap { bitmap: Bitmap? ->
+                    if (bitmap == null) {
+                        return@flatMap Single.error<Bitmap>(Throwable("Error occurred while applying paint"))
+                    } else {
+                        return@flatMap Single.just(bitmap)
+                    }
```

Comment:
this seems to terminate the flatmap early in every case

Reply:
This seems to be working fine. We had to return the bitmap or the error